### ISSUE TITLE
Cryptpad won't load in Firefox 99

### ISSUE THREAD

**Issue by kelimuttu** [mozilla-mobile/firefox-ios#10463]
## Steps to reproduce
* Update to Firefox 99
* Go to https://cryptpad.fr/drive/

### Expected behavior
Page will show up

### Actual behavior
Get a "Script error" message instead

### Device & build information
* Device: iOS 15.3.1
* Build version: Firefox 99
* 
### Notes
It works well previously with Firefox 98.2. It also works in Safari.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-4085)

**Comment by lmarceau** [comment]
@kelimuttu Hello! Out of curiosity has this problem been raised at the Cryptpad people as well? https://github.com/xwiki-labs/cryptpad

**Comment by kelimuttu** [comment]
@lmarceau not sure about that. The only similar issue I could find is [this one](https://github.com/xwiki-labs/cryptpad/issues/884), but it's a desktop issue so I suppose it's different. 

This was originally reported through [Twitter](https://twitter.com/tobiaalberti/status/1512102097390604293) and I was able to reproduce the issue upon upgrading to 99.

**Comment by ansuz** [comment]
> @kelimuttu Hello! Out of curiosity has this problem been raised at the Cryptpad people as well? https://github.com/xwiki-labs/cryptpad

It was reported but it seems to only occur on Firefox 99 for iOS, and unfortunately nobody on our team has the necessary hardware to access what is triggering the script error.

If anybody can provide a stack trace we should be able to guard against it on our side. That said, it seems like that would fix the symptom and not the cause.

**Comment by ansuz** [comment]
> I'm trying to understand, what is the specific need? I see the error page in a simulator as well, so I don't think hardware is needed?

Can you recommend a good simulator? I've tried BrowserStack (which seems to only simulate Safari on iOS) and Lambdatest  (which simulates Firefox on iOS but doesn't reproduce this behaviour). It seems reliable on a physical iPhone running the latest version.

**Comment by lmarceau** [comment]
By simulator I meant running from Xcode on a Simulator iPhone device. From there you can use Safari web inspector to have the Console and see some logs. Hope this helps!

![Screen Shot 2022-04-14 at 2 16 48 PM](https://user-images.githubusercontent.com/11338480/163452195-fe2a21ab-0e86-4046-a52f-b1367a9c2b9f.png)

![simulator_screenshot_7C929820-B5D2-4854-B10C-C1E3ABCCB1A9](https://user-images.githubusercontent.com/11338480/163452170-c8f340b3-dd1e-4429-91e8-b80c6458268e.png)

**Comment by ansuz** [comment]
> By simulator I meant running from Xcode on a Simulator iPhone device. From there you can use Safari web inspector to have the Console and see some logs. Hope this helps!

Unfortunately I don't see anything in the screenshot that could explain a script error.

Perhaps I'm wrong about this, but my understanding is that Xcode will only run on Apple devices. Noone on our team has a macbook or similar device, so it looks like we're out of options for debugging this further.

**Comment by mthld** [comment]
I recently tried on the latest Firefox Daylight v103.1 (15076) and it works as intended with your enhanced tracking protection disabled for cryptpad.fr.

@kelimuttu could you confirm if it's now working on your end?

**Comment by kelimuttu** [comment]
@mthld Nope. It still gave me "script error" on Daylight 103.1.

**Comment by mthld** [comment]
Even with enhanced tracking disabled? It's working for me right now. :o

**Comment by mthld** [comment]
> @mthld Nope. It still gave me "script error" on Daylight 103.1. 

Never mind, I tried again today and I can't reproduce it. I wonder what I've done past days... 🤦🏻‍♀️

**Comment by mthld** [comment]
> By simulator I meant running from Xcode on a Simulator iPhone device. From there you can use Safari web inspector to have the Console and see some logs. Hope this helps!

Hi @lmarceau,

Could you share some insights regarding this debugging setup? As I now possess a MacBook I would like to reproduce this issue.

I also regret that [the debugging option](https://support.mozilla.org/en-US/questions/1270192#answer-1258417) was removed from Firefox iOS, as [this very convenient tool](https://support.mozilla.org/en-US/kb/report-site-issues-firefox-ios) isn't available anymore too.

**Comment by lmarceau** [comment]
Report site issue is available in Beta and Developer builds, so if you build from Xcode you'll get this menu option. As for debug options, there's some that are available if you press the build version in the settings 5 times (available on all build types).

As for the step to reproduce what I explained there, first you need to build on Xcode (this can be done following the readme instructions). Once you have that, you need the Safari web inspector. You can enable it [here](https://developer.apple.com/library/archive/documentation/NetworkingInternetWeb/Conceptual/Web_Inspector_Tutorial/EnableWebInspector/EnableWebInspector.html). You can then debug the iPhone session following the instructions [here](https://www.idownloadblog.com/2019/06/21/how-to-use-safari-web-inspector-ios-mac/). Let me know if more details is needed.

**Comment by mthld** [comment]
Thanks for the explanations!

I've encountered some dependencies issues but worked my way around. However I'm now facing another problem, where all content-blocker lists seems missing:
```
error: /Users/mathilde/Code/cryptpad/firefox-ios/content-blocker-lib-ios/Lists/disconnect-block-cryptomining.json: No such file or directory (in target 'Client' from project 'Client')
```

<img width="500" alt="Screenshot 2022-08-22 at 10 08 07" src="https://user-images.githubusercontent.com/41020854/185871632-f6d25867-01f0-455d-a110-9045e431b5f8.png">

**Comment by mthld** [comment]
Thanks for the explanations!

I encountered some dependencies issues but figured out how to manage that. However I'm now facing another problem with content-blocker lists:

```
error: /Users/mathilde/Code/cryptpad/firefox-ios/content-blocker-lib-ios/Lists/disconnect-block-social.json: No such file or directory (in target 'Client' from project 'Client')
```

<img width="465" alt="Screenshot 2022-08-22 at 10 38 11" src="https://user-images.githubusercontent.com/41020854/185877831-ab54b51c-a6e4-4d0b-84e7-aa4b92f56f34.png">

**Comment by mthld** [comment]
Hello, thanks for the explanation @lmarceau!

I successfully built and ran Firefox iOS in a simulator. Here are the 2 errors found:
```
[Error] TypeError: undefined is not an object (evaluating 'e.replace')
	r (user-script:7:1:36378)
	(anonymous function) (user-script:7:1:37848)
```

```
[Error] Script error.
	(anonymous function) (boot2.js:77)
```

@ansuz does that ring any bell from our side?

**Comment by mthld** [comment]
If needed here is the whole file with the errors at lines `1176` & `1217`: https://gist.github.com/mthld/61bb4a6c9c680b4bbe1bce4eb1145e04

**Comment by ansuz** [comment]
Hi!

From reading the code @mthld linked, it seems the error is related to Firefox's [Reader View](https://support.mozilla.org/en-US/kb/view-web-pages-reader-view-firefox-ios).

The inline `r` function assumes its `e` argument is a string, but that doesn't seem guaranteed. I can propose this minimal patch to guard against the resulting type error:

```diff
             function r(e) {
+                if (typeof(e) !== 'string') { return ''; }
                 return e.replace(/\&/g, "&amp;").replace(/\</g, "&lt;").replace(/\>/g, "&gt;").replace(/\"/g, "&quot;").replace(/\'/g, "&#039;")
             }
```

...but I'm not sure where I could submit a PR.

**Comment by lmarceau** [comment]
@ansuz This code seems to be from the `function escapeHTML(string) {` in `ReaderMode.js`. I tested and this seems to fix the problem 🎉 Let me know if you want to open the PR, as this is your fix.

**Comment by ansuz** [comment]
_Submitted_

If anybody needs me I'll be updating my CV to include _senior iOS developer_ ;)

Thanks for your help!

**Comment by data-sync-user** [comment]
➤ Laurie Marceau commented:

For QA we can also test that reader mode works as expected, no changes.

**Comment by data-sync-user** [comment]
➤ Simion Basca commented:

Verifying as fix on main 59568c9ea028c11959f8b02501c6503372385047. Reader mode has correct functionality and  https://cryptpad.fr/drive/ ( https://cryptpad.fr/drive/|smart-link ) loads correctly.

**Comment by data-sync-user** [comment]
➤ Simion Basca commented:

Verified fix on v106 (19594)
